### PR TITLE
Search config

### PR DESCRIPTION
**Pull request by kaapstorm**
This is the project settings page.

Module settings to follow.

@proteusvacuum @snopoke 

I have an observableArray that [doesn't seem to be observing](https://github.com/dimagi/commcare-hq/blob/c25faab08921e23b401908b9dca305b72ca39263/corehq/apps/domain/static/domain/js/case-search-config.js#L33-L33) very well. Changing values in the UI does not change values in the viewModel. I must be missing something.

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 10 errors:

* corehq/apps/domain/static/domain/js/case-search-config.js, line 33 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 59 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 46 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/views.py, line 2108 - E702 multiple statements on one line (semicolon)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 13 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 16 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 49 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 62 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 60 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/views.py, line 14 - F401 'UpdateView' imported but unused

**Comment by czue** [comment]
@kaapstorm :computer: :sunrise: :zzz:

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/domain/static/domain/js/case-search-config.js, line 33 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 59 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 46 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 13 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 16 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 49 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 62 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 60 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/views.py, line 14 - F401 'UpdateView' imported but unused

**Comment by dimagimon** [comment]
There are 8 errors:

* corehq/apps/domain/static/domain/js/case-search-config.js, line 33 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 59 - Missing trailing comma. (comma-dangle)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 46 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 13 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 16 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 49 - 'event' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 62 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/domain/static/domain/js/case-search-config.js, line 60 - Missing trailing comma. (comma-dangle)

File: corehq/apps/domain/static/domain/js/case-search-config.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self.removeProperty = function (property) {
+            self.properties.remove(property);
+        };
+    };
+
+    /**
+     * Returns a viewModel for domain/admin/case_search.html
+     */
+    module.CaseSearchConfig = function (options) {
+        var self = this;
```
Comment (dimagimon) [inline corehq/apps/domain/static/domain/js/case-search-config.js]: 'self' is defined but never used (no-unused-vars)

**Comment by kaapstorm** [comment]
cc @millerdev

File: corehq/apps/domain/static/domain/js/case-search-config.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+hqDefine('domain/js/case-search-config.js', function () {
+    'use strict';
+
+    var module = {};
+
+    var CaseTypeProps = function (caseType, properties) {
+        var self = this;
+        self.case_type = ko.observable(caseType);
+        self.properties = ko.observableArray(properties);
+        // TODO: ^^^ This observableArray isn't changing the values of the strings inside it.
```
Comment (proteusvacuum) [inline corehq/apps/domain/static/domain/js/case-search-config.js]: I don't think you can just change strings in an observableArray. From the knockout docs:
> Key point: An observableArray tracks which objects are in the array, not the state of those objects

File: corehq/apps/domain/static/domain/js/case-search-config.js (hunk 1 above)
Comment (proteusvacuum) [inline corehq/apps/domain/static/domain/js/case-search-config.js]: you are lying, dimagimon.

File: corehq/apps/domain/views.py [hunk 3]
```diff
@@ -2071,6 +2080,57 @@ def post(self, request, *args, **kwargs):
...
+    template_name = 'domain/admin/case_search.html'
+
+    @method_decorator(domain_admin_required)
+    @use_bootstrap3
+    def dispatch(self, request, *args, **kwargs):
+        return super(CaseSearchConfigView, self).dispatch(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+
+        def unpack_fuzzies(query_dict):
```
Comment (proteusvacuum) [inline corehq/apps/domain/views.py]: yipes! I'm guessing my helper methods on CaseSearchConfig were not that helpful?

File: corehq/apps/domain/views.py [hunk 4]
```diff
@@ -2071,6 +2080,57 @@ def post(self, request, *args, **kwargs):
...
+            for key, value in query_dict.items():
+                match = pattern.match(key)
+                if match:  # non-fuzzy-properties won't match, i.e. "^enable$"
+                    fuzzy_dict[int(match.group('index'))][match.group('attr')] = value
+            if not fuzzy_dict:
+                return []
+            return [fuzzy_dict[i] for i in range(max(fuzzy_dict.keys()) + 1) if fuzzy_dict[i]]
+
+        try:
+            case_search_config = CaseSearchConfig.objects.get(pk=self.domain)
```
Comment (proteusvacuum) [inline corehq/apps/domain/views.py]: I think you could use `CaseSearchConfig.objects.get_or_create` here. or does that save it immediately, which is why you did it this way?

**Comment by proteusvacuum** [comment]
Looking good Norman!

File: corehq/apps/app_manager/static/app_manager/js/detail-screen-config.js [hunk 5]
```diff
@@ -196,6 +196,38 @@ hqDefine('app_manager/js/detail-screen-config.js', function () {
...
     };
 
+    var searchViewModel = function (searchProperties, saveButton) {
+        var self = this;
+
+        var SearchProperty = function (name, label) {
+            var self = this;
+            self.name = ko.observable(name);
+            self.label = ko.observable(label);
+        }
```
Comment (dimagimon) [inline corehq/apps/app_manager/static/app_manager/js/detail-screen-config.js]: Missing semicolon. (semi)

File: corehq/apps/app_manager/views/modules.py [hunk 6]
```diff
@@ -167,7 +169,8 @@ def _get_basic_module_view_context(app, module):
         'case_list_form_not_allowed_reason': allow_case_list_form,
         'valid_parent_modules': _get_valid_parent_modules(app, module),
         'child_module_enabled': toggles.BASIC_CHILD_MODULE.enabled(app.domain)
-                                 and module.doc_type != "ShadowModule"
+                                 and module.doc_type != "ShadowModule",
```
Comment (dimagimon) [inline corehq/apps/app_manager/views/modules.py]: E131 continuation line unaligned for hanging indent

File: corehq/apps/domain/static/domain/js/case-search-config.js (hunk 2 above)
Comment (kaapstorm) [inline corehq/apps/domain/static/domain/js/case-search-config.js]: :tada: Aha! **Key point**! Indeed. Thank you.

**Comment by kaapstorm** [comment]
Added module UI stuff. View code is next up.

**Comment by proteusvacuum** [comment]
I think you might have to reopen these PR's against master so that travis runs.

**Comment by kaapstorm** [comment]
Hmmm. I also need this stuff in the other branch. Closing this PR, merging this branch into suite_xml and deleting.

**Pull request closed without merging**